c10/cuda/CUDAMallocAsyncAllocator.cpp

```diff
@@ -0,0 +1,823 @@
+  TORCH_INTERNAL_ASSERT(
+      0 <= device && static_cast<size_t>(device) < device_count,
+      "Invalid device index ",
+      device,
+      ": did you call init?");
+
+  // If stream is a null (default) stream,
+  // cudaMallocAsync infers the device from the ambient context,
+  // so we need to set the right ambient context.
+  CUDAGuard g(device);
+
+  std::lock_guard<std::mutex> lk(general_mutex);
+
+  lazy_init_device(device);
+
+  // Defensively checks for preexisting CUDA error state.
+  auto err = cudaGetLastError();
+  C10_CUDA_CHECK(err);
+
+  // TODO: Could we avoid calling cudaMallocAsync while holding general_mutex,
+  // perhaps by letting lazy_init_device use separate once_flags or an internal
+  // static initializer?
+  if (pytorch_used_bytes[device] + size > pytorch_memory_limits[device]) {
+    err = cudaErrorMemoryAllocation;
+  } else {
+    err = cudaMallocAsync(devPtr, size, stream);
+  }
+
+  if (err == cudaErrorMemoryAllocation) {
+    // Clears CUDA's internal error state so the user, if desired, can catch the OOM
+    // exception, free some stuff on the script side, and retry the allocation.
+    // This aligns with the behavior of alloc_block in CUDACachingAllocator.cpp.
+    cudaGetLastError();
+    size_t device_free;
+    size_t device_total;
+    C10_CUDA_CHECK(cudaMemGetInfo(&device_free, &device_total));
+    TORCH_CHECK_WITH(CUDAOutOfMemoryError,
+                     false,
+                     "Allocation on device ", device, " would exceed allowed memory.",
+                     "\nCurrently allocated     : ", format_size(pytorch_used_bytes[device]),
+                     "\nRequested               : ", format_size(size),
+                     "\nDevice limit            : ", format_size(device_total),
+                     "\nFree (according to CUDA): ", format_size(device_free),
+                     "\nPytorch limit (set by user-supplied memory fraction)"
+                     "\n                        : ", format_size(pytorch_memory_limits[device]));
+  } else {
+    C10_CUDA_CHECK(err);
+  }
+
+  auto inserted = ptr_info.emplace(*devPtr, PtrUsage(size, capture_underway));
+  TORCH_INTERNAL_ASSERT(inserted.second,
+                        "address returned by cudaMallocAsync already exists "
+                        "in ptr_info");
+
+  inserted.first->second.usage_streams.emplace_back(stream, device);
+
+  pytorch_used_bytes[device] += size;
+}
+
+} // anonymous namespace
+
+// Same pattern as CUDACachingAllocator.cpp.
+// Again, I don't think we absolutely need the symmetry,
+// but it's the simplest way to imitate the interface.
+struct CudaCachingAllocator : public Allocator {
+  DataPtr allocate(size_t size) const override {
+    constexpr size_t one_exa_bytes = 1152921504606846976ULL;
+    TORCH_CHECK_WITH(
+        CUDAOutOfMemoryError,
+        size < one_exa_bytes,
+        "CUDA out of memory. Tried to allocate more than 1EB memory.");
```

Comment:
Do you remember why we added this check?

Reply:
No idea. It's something i'm imitating from the native allocator, and was present there before I ever looked at it.

Reply:
To avoid internal assert on oversized allocations #57571